### ISSUE TITLE
Cypress can not see the visable elements in a overflow scroll div

### ISSUE THREAD

**Issue by scottdavis** [cypress-io/cypress#57]
Trying to grab an element in an overflow scroll-x div to click on it but keep getting the error that the element is not visible even though I can clearly see it in the test runner window.

**Comment by brian-mann** [comment]
Does the `<div>` itself displace dimensions? I've seen this situation before where the parent container actually has `0` width or `0` height because it's children are floated. In that case, the element *is* technically hidden even though it's children are visible.

There are some better error messages coming in the next release or two that specify the exact reason why Cypress thinks the element isn't visible.

Additionally you can always force Cypress to interact with the element regardless of its visibility by passing `{force: true}` to the command.

If this isn't the case however then this is a bug.

**Comment by scottdavis** [comment]
It is this case: 
"I've seen this situation before where the parent container actually has 0 width or 0 height because it's children are floated. In that case, the element is technically hidden even though it's children are visible.
"

**Comment by scottdavis** [comment]
Some Context (if that helps): The test case I'm trying to create is click on an image in our slide bar and having the bar scroll to n location

**Comment by scottdavis** [comment]
figured it out yay!

**Comment by brian-mann** [comment]
What was the problem, haha.

**Comment by scottdavis** [comment]
Using force allowed me to click the div and I was able to get a handle to the scroll property with chai jquery

**Comment by scottdavis** [comment]
``` coffeescript
  it "should scroll by control", ->
    cy.visit '/?autoplay=false'
    cy
      .get('.thumbs .scroll').should("have.prop", 'scrollLeft').and('match', /^0$/)
      .get('.thumb:nth-child(3) photo').click force: true
      .wait(800)
      .get('.thumbs .scroll').should("have.prop", 'scrollLeft').and('match', /^5\d\d$/)
      .get('.thumbs .right-control').click force: true
      .wait(800)
      .get('.thumbs .scroll').should("have.prop", 'scrollLeft').and('match', /^7\d\d$/)
      .get('.thumbs .left-control').click force: true
      .wait(800)
      .get('.thumbs .scroll').should("have.prop", 'scrollLeft').and('match', /^0$/)

```

**Comment by scottdavis** [comment]
was an adventure in sub pixel rounding and css animations but i got it locked down

**Comment by scottdavis** [comment]
But it maybe nice to have some way to get a handle to a dom element so I could get its top or left property

**Comment by brian-mann** [comment]
[You can, it's called aliasing](https://github.com/cypress-io/cypress/wiki/aliasing).

Additionally you can use chai's [http://chaijs.com/api/bdd/](within) chainer to avoid those regex's.

Also I see why you're using the `wait`. In the next release Cypress will be able to predict assertions coming downstream which should alleviate the need to write those in this case.

**Comment by scottdavis** [comment]
the waits are waiting on the css animation to finish playing not sure if i needed to do that.

**Comment by scottdavis** [comment]
Ah i get aliases now! On that note is there a way to bundle common assertions into macros?

**Comment by brian-mann** [comment]
You can create custom commands, that part is not documented but in fact all the commands in Cypress are already built with this, and there is an example of custom commands here: 

https://github.com/cypress-io/examples-react-travis-ci/blob/master/tests/_support/spec_helper.js

However I'm not sure what you'd want to do as far as bundling common assertions into macros. Maybe you can give me an example.

**Comment by scottdavis** [comment]
``` coffeescript
describe "Owl Test", ->
  beforeEach ->
     cy.visit '/'
  describe "highlighter", ->
    SCROLL_OFFSET = 112
    navScrollMacro = (navText, elementId, offset=SCROLL_OFFSET) ->
      cy
        .contains('.sticky-nav a', navText)
        .click()
        .wait(1200)
        .contains('.sticky-nav a', navText).parent().should('have.class', 'active')
        .get(elementId).then (ele) ->
          cy
            .get('body').should('have.prop', 'scrollTop').and('equal', ele[0].offsetTop - offset)

    it "should scroll to agent", ->
      navScrollMacro('Agent', '#agent')
    it "should scroll to map", ->
      navScrollMacro('Map', '#map')
    it "should scroll to share", ->
      navScrollMacro('Share', '#share')
    it "should scroll to gallery", ->
      navScrollMacro('Gallery', '#gallery')
    it "should scroll to description", ->
      # Description ahs a 5 px padding
      navScrollMacro('Description', '#description', SCROLL_OFFSET - 3)
```

Test is a little weird i usually wouldn't be testing the browser's ability to just to an anchor but there is a jquery plugin around the smooth scrolling so i just turned the repeated code into a function to dry up the test case.

**Comment by brian-mann** [comment]
Right so I'm not going to debate the validity of the test - but assuming this is what you'd like to test, then turning this into a custom command is exactly why they were created. The example I pasted above shows this happening.

You can look at https://github.com/cypress-io/examples-react-travis-ci/blob/master/tests/app_spec.js to see how those custom commands are consumed.

So just create a custom command in your `_support` folder and you can use it with `cy`. You can also determine whether you want to silence the log output from the individual commands, or wrap those up into one specific custom command. You can control everything from the arguments to the output to the Dev Tools console.

You'd want to move the `SCROLL_OFFSET` constant into the custom command function and then just provide a delta `Number` as the 3rd argument. So just pass `-3` in the "Description" example.

Play around with it and let me know what you come up with.

Also you can remove the 2nd `contains` because `wait` will just return the same subject as what was passed into it, which is your `<a>`. There's no need to go find it again.

**Comment by scottdavis** [comment]
Wow thats awesome :clap: Thanks so much
